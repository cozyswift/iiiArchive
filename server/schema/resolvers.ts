import { DateTimeResolver, URLResolver } from 'graphql-scalars';
import {
  materialList,
  imgList,
  Material,
  ImageArr,
  MaterialTypes,
  Keyword,
} from '../db';
import { Resolvers } from '../types/graphql';
// import {Resolvers}from '../types/graphql'
const resolvers: Resolvers = {
  DateTime: DateTimeResolver,
  URL: URLResolver,

  Material: {
    // materialType(typeId: any) {
    //   return materialTypes.find(mTypes => mTypes.typeName === 'jpg');
    // },

    picture(root) {
      const result = imgList.filter(img => img.materialId == root.id) || null;

      return result;
    },
  },

  // Materials: {
  //   material(findMaterial: any) {
  //     return material.find(m => m.id === findMaterial.id);
  //   },
  // },

  Query: {
    materialList() {
      return materialList;
    },

    material(root, { materialId }) {
      const result = materialList.find(m => m.id === materialId);

      return result || null;
    },

    // archivistID({ archivistId }: any){
    //   const findMaterial=materials.find(mts=>mts.id===archivistId);
    //   if(findMaterial && findMaterial.archivist){
    //     const result=material.find(material=>material.archivistId===findMaterial.archivist);
    //     const archivist=result?.archivistId;
    //     return archivist;
    //   }
    // }
  },

  Mutation: {
    addMaterial(root: any, { title, archivistId }: any) {
      console.log('root', { root });
      console.log('title', { title });
      console.log('archivis', { archivistId });

      const newMaterialId = (materialList.length + 1).toString();
      console.log({ newMaterialId });
      const newArchivistId = archivistId;
      const newDescription = 'test';
      const imgArr = {
        id: (imgList.length + 1).toString(),
        regUser: archivistId,
        fileName: '네임 테스트',
        examUrl: [
          'https://images.unsplash.com/photo-1461360370896-922624d12aa1?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=753&q=80%22',
        ],
        width: '',
        height: '',
        tag: ['tag'],
        regIp: '',
        fileType: 'jpg',
        materialId: newMaterialId,
      };

      const newCreatedAt = new Date();
      const newCreator = 'test';
      const newMaterialType: any = 'jpg';
      const newDonor: [] = [];
      const newKeyword: [] = [];
      const newIdentificationNum: [] = [];
      const newTitle = title;

      imgList.push(imgArr);

      const newMaterial = {
        id: newMaterialId,
        archivistId: newArchivistId,
        title: newTitle,
        description: newDescription,
        picture: [imgList.find(m => m.materialId === newMaterialId)],
        createdAt: newCreatedAt,
        creator: newCreator,
        materialType: newMaterialType,
        donor: newKeyword,
        keyword: newKeyword,
        identificationNum: newIdentificationNum,
      };

      materialList.push(newMaterial);

      // console.log({imageArr})
      // console.log({material});

      const result = materialList.map(m => {
        materialList.find(m => {
          m.archivistId === archivistId;
        });
      });

      // console.log(material.find(m => m.archivistId === archivistId));

      return newMaterial;
    },

    addNewMaterial(root, { newMaterial }:any, { pubsub }) {
      console.log({ root });
      console.log({ newMaterial });

      materialList.push(newMaterial);

      // pubsub.publish('materialAdded', {
      //   materialAdded: newMaterial,
      // });

      return materialList[0];
    },
  },

  Subscription: {
    materialAdded: {
      subscribe: (root, args, { pubsub }) =>
        pubsub.asyncIterator('materialAdded'),
    },
  },
};

export default resolvers;
