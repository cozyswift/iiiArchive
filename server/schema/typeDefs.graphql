scalar DateTime
scalar URL


type MaterialType {
    id:ID!
    typeName:String!
}

input InputMaterialType {
    id:ID!
    typeName:String!
}


type ImageArr {
    id:ID!
    regUser:String!
    fileName:String!
    examUrl:[URL!]
    width:String
    height:String
    tag:[String]
    regIP:String
    fileType:String
    materialId:String
}



input InputImageArr {
    id:ID!
    regUser:String!
    fileName:String!
    examUrl:[URL!]
    width:String
    height:String
    tag:[String]
    regIP:String
    fileType:String
    materialId:String
}

type Material {
    id:ID!
    archivistId:String!
    title:String!
    description:String
    picture:[ImageArr]!
    createdAt:DateTime!
    creator:String
    materialType:MaterialType
    donor:[String!]
    keyword:[String]
    identificationNum:[String]
}

input InputMaterial {
    id:ID
    archivistId:String
    title:String
    description:String
    picture:[InputImageArr]
    createdAt:DateTime
    creator:String
    materialType:InputMaterialType
    donor:[String!]
    keyword:[String]
    identificationNum:[String]
}


type Archivist {
    id:ID!
    archivistId:ID!
    name:String!
    material:[Material!]
}

type Query {
    materialList:[Material]
    material(materialId:ID!):Material
    archivist:Archivist
}

type Mutation {
    addMaterial(title:String,archivistId:String):Material
    addNewMaterial(newMaterial:InputMaterial):Material
}
