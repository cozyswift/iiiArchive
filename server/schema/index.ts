import { importSchema } from 'graphql-import';
import { makeExecutableSchema } from 'graphql-tools';
import resolvers from './resolvers';
// const typeDef = importSchema('schema/typeDefs.graphql');

const typeDefs=`
scalar Date
scalar URL


type Keywords {
    id:ID!
    keyword:String
}

type MaterialTypes {
    id:ID!
    typeName:String
}


type Material {
    id:ID!
    title:String
    description:String
    picture:URL
    createdAt:Date
    materialType:MaterialTypes
    donor:[String]
    keyword:[String]
    identificationNum:[String]
}

type Materials {
    id:ID!
    archivist:String
    material:Material
}

type Query {
    materials:[Materials]
}


`;



export default makeExecutableSchema({ resolvers, typeDefs});
