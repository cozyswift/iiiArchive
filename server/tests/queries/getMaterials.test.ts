import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server-express';
import schema from '../../schema';

describe('머테리얼 쿼리', () => {
  it('모든 사료데이터를 가져와야 한다', async () => {
    const server: any = new ApolloServer({ schema });
    //스키마를 주입하여 가짜 GraphQl서버 생성

    const { query } = createTestClient(server);

    const res = await query({
      query: gql`
        query GetMaterials {
          materials {
            id
            archivist
            material {
              id
              title
              description
              picture
              createdAt
              keyword
              materialType {
                id
                typeName
              }
              identificationNum
            }
          }
        }
      `,
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
