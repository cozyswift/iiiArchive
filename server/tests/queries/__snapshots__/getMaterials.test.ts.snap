// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`머테리얼 쿼리 모든 사료데이터를 가져와야 한다 1`] = `
Object {
  "materials": Array [
    Object {
      "archivist": "주강민",
      "id": "1",
      "material": Object {
        "createdAt": "2020-03-04T16:47:48.125Z",
        "description": "박근혜는 하야하라",
        "id": "1",
        "identificationNum": Array [
          "ad003-A00008",
        ],
        "keyword": Array [
          "서면",
          "박근혜퇴진본부",
          "하야",
          "퇴진",
        ],
        "materialType": Object {
          "id": "1",
          "typeName": "jpg",
        },
        "picture": "http://iiiarchives.org/wp-content/uploads/2018/11/ad003-A-00008-min-1024x205.jpg",
        "title": "현수막(4)",
      },
    },
    Object {
      "archivist": "조수정",
      "id": "2",
      "material": Object {
        "createdAt": "2020-03-04T16:47:48.125Z",
        "description": "박근혜 파면이 결정된 3월 10일의 131차 시국집회의 모습이다.",
        "id": "2",
        "identificationNum": Array [
          "pi003-A-00275",
          "pi003-A-00276",
          "pi003-A-00277",
          "pi003-A-00278",
          "pi003-A-00279",
          "pi003-B-00099",
          "pi003-B-00100",
          "pi003-B-00101",
          "pi003-B-00102",
          "pi003-B-00103",
          "pi003-B-00104",
          "pi003-B-00105",
        ],
        "keyword": Array [
          "서면",
          "시국집회",
          "탄핵인용",
          "촛불승리",
          "거리행진",
        ],
        "materialType": Object {
          "id": "1",
          "typeName": "jpg",
        },
        "picture": "http://iiiarchives.org/wp-content/uploads/collection/picture/003/170310%20%EC%B4%9B%EB%B6%88%EC%8A%B9%EB%A6%AC%20%EC%8B%9C%EA%B5%AD%EC%A7%91%ED%9A%8C/pi003-B-00103-1024x683.jpg",
        "title": "170310 촛불승리 시국집회",
      },
    },
    Object {
      "archivist": "주강수",
      "id": "3",
      "material": Object {
        "createdAt": "2020-03-04T16:47:48.125Z",
        "description": "1월 25일 부산시국집회의 모습이다.",
        "id": "3",
        "identificationNum": null,
        "keyword": Array [
          "서면",
          "시국집회",
          "천주교단체",
          "거리행진",
        ],
        "materialType": Object {
          "id": "1",
          "typeName": "jpg",
        },
        "picture": "http://iiiarchives.org/wp-content/uploads/collection/picture/003/170125%20%EB%B9%84%EC%A3%BC%EB%A5%98%EC%82%AC%EC%A7%84%EA%B4%80%20%EC%B2%9C%EC%A3%BC%EA%B5%90%EB%8B%A8%EC%B2%B4%20%EC%8B%9C%EA%B5%AD%EB%8C%80%ED%9A%8C/pi003-B-00081-1024x683.jpg",
        "title": "170125 천주교단체 시국집회",
      },
    },
  ],
}
`;
